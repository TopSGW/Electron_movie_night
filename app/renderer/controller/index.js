// Logic for mapping application state onto component props.
import {
  App,
  DisplayMoviesContainer,
  ImportMovies,
  ImportStats,
  SearchMovies,
  MainContent,
  MovieThumbnailContainer
} from './containers'

// Logic for mapping application events directly to redux dispatch actions.
import './mapEventsToDispatch'

// Re-export all the containers.
export {
  App,
  DisplayMoviesContainer,
  ImportMovies,
  ImportStats,
  MainContent,
  MovieThumbnailContainer,
  SearchMovies
}
