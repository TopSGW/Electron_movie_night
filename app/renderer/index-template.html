<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <title>Movie Night</title>
    <link rel="stylesheet" href="../vendor/font-awesome-4.7.0/css/font-awesome.min.css">
    <%=htmlWebpackPlugin.options.css%>
  </head>
  <body style="background-color:rgb(20,20,20) !important; /* override bootstrap */">
    <div id="root"></div>
    <script src="<%=htmlWebpackPlugin.options.bundle%>"></script>
  </body>
</html>
