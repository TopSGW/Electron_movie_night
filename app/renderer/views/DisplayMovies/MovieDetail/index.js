import fs from 'fs'
import React, { Component } from 'react'
import styled from 'styled-components'
import { shell } from 'electron'

import { fileExists, filePathToUrl } from '../../../../shared/utils'
import { fetchMovieMetadata, fetchMovieMetadataByID } from '../../../../background/api/fetchMovieMetadata'
import { Close, Edit, Trash } from '../../../icons'

import ListMeta from './ListMeta'
import Location from './Location'
import Meta from './Meta'
import PlayMovieButton from './PlayMovieButton'
import Ratings from './Ratings'
import Title from './Title'
import Update from './Update'

export default class MovieDetail extends Component {
  constructor (props) {
    super(props)

    this.openMovieInDefaultPlayer = this.openMovieInDefaultPlayer.bind(this)
    this.showMovieInFinder = this.showMovieInFinder.bind(this)
    this.updateFileAvailable = this.updateFileAvailable.bind(this)
    this.close = this.close.bind(this)
    this.handleRedoSearch = this.handleRedoSearch.bind(this)
    this.handleSaveSearch = this.handleSaveSearch.bind(this)
    this.handleMoveToTrash = this.handleMoveToTrash.bind(this)
    this.handleToggleEdit = this.handleToggleEdit.bind(this)
    this.anchor = null

    this.state = {
      fileAvailable: true,
      editOpen: false,
      searching: false,
      searchError: '',
      searchMovieResult: null,

      // Shadow incoming movie prop.
      // This is so after making edits we can save
      // results local to component w/o impacting redux state.
      movie: props.movie
    }
  }

  componentDidMount () {
    this.props.center &&
    this.anchor &&
    this.anchor.scrollIntoViewIfNeeded() // centers anchor in viewport

    this.updateFileAvailable()
    fs.watchFile(this.state.movie.location, this.updateFileAvailable)
  }

  componentWillUnmount () {
    fs.unwatchFile(this.state.movie.location, this.updateFileAvailable)
  }

  updateFileAvailable () {
    fileExists(this.state.movie.location)
      .then(result => this.setState({ fileAvailable: result }))
  }

  openMovieInDefaultPlayer (e) {
    e.preventDefault()
    const { movie } = this.state
    shell.openItem(movie.location)
  }

  showMovieInFinder (e) {
    e.preventDefault()
    const { movie } = this.state

    shell.showItemInFolder(movie.location)
  }

  close () {
    const { handleCloseMovieDetails } = this.props
    const { movie } = this.state
    if (handleCloseMovieDetails) {
      handleCloseMovieDetails(movie)
    }
  }

  handleRedoSearch (searchTitle, searchYear, imdbID) {
    const { movie } = this.state
    const searchQuery = imdbID || (searchTitle + (searchYear ? ` [${searchYear}]` : ''))

    this.setState({ searching: true, searchError: '' })

    const apiCall = imdbID ? fetchMovieMetadataByID : fetchMovieMetadata
    apiCall(searchQuery)
      .then(response => {
        const searchMovieResult = {
          ...response,
          location: movie.location,
          query: `&s=${searchTitle}&y=${searchYear}`,
          fileSize: movie.fileSize
        }

        this.setState({
          searching: false,
          searchMovieResult,
          searchError: ''
        })
      })
      .catch(error => {
        console.error(error)

        this.setState({
          searching: false,
          searchMovieResult: null,
          searchError: `Search Error: Sorry no result :-(`
        })
      })
  }

  handleSaveSearch () {
    const { onUpdateMovieMetadata } = this.props
    const { searchMovieResult } = this.state

    if (searchMovieResult) {
      onUpdateMovieMetadata(searchMovieResult)
      this.setState({
        editOpen: false,
        movie: searchMovieResult,
        searchMovieResult: null
      })
    }
  }

  handleMoveToTrash () {
    const { onMoveToTrash } = this.props
    const { movie } = this.state
    const response = window.confirm(`Warning: this will DELETE the following file from your hard drive and move it to the trash:\n\n${movie.location}`)
    if (response) {
      if (
        shell.moveItemToTrash(movie.location) ||
        window.confirm(`Permission denied. Cannot move file to trash\n\nDelete it from your Movie Night database?`)
      ) {
        onMoveToTrash(movie)
        this.close()
      }
    }
  }

  handleToggleEdit () {
    const { editOpen, searchMovieResult } = this.state

    if (editOpen && searchMovieResult) {
      const discard = window.confirm('Discard All Edits?')
      if (!discard) { return }
      this.setState({ searchMovieResult: null })
    }

    this.setState({ editOpen: !editOpen })
  }

  render () {
    const { allowEdit } = this.props
    const { movie: dbMovie } = this.state
    const { editOpen, searching, searchError, searchMovieResult } = this.state

    const movie = searchMovieResult || dbMovie

    const posterUrl = searchMovieResult
      ? searchMovieResult.imgUrl
      : (movie.imgFile ? filePathToUrl(movie.imgFile) : '')

    const Edit = editOpen ? EditButtonRed : EditButton

    return (
      <FlexboxDiv >
        <button /* Hook for scrolling top of div to center of viewport */
          ref={(node) => { if (node) this.anchor = node }}
          style={{visibility: 'hidden'}}
        />

        <Poster fileUrl={posterUrl}>
          {this.state.fileAvailable && <PlayMovieButton onClick={this.openMovieInDefaultPlayer} />}
        </Poster>

        <MovieDetailsContainer>
          <header>
            <div style={{display: 'flex', flexDirection: 'row', justifyContent: 'flex-end', width: '100%', marginBottom: '8px'}}>
              {allowEdit && <Edit onClick={this.handleToggleEdit} />}
              <TrashButton onClick={this.handleMoveToTrash} />
              <CloseButton onClick={this.close} />
            </div>
            <Title title={movie.title} editing={editOpen} />
            <Meta year={movie.year} rated={movie.rated} runtime={movie.runtime} />
            <Ratings
              audience={Number(movie.imdbRating) / 10.0}
              critics={Number(movie.metascore) / 100.0}
            />
          </header>

          <Spacer small={editOpen} />

          <VerticalScrollSection>
            {movie.plot}
          </VerticalScrollSection>

          <Spacer small={editOpen} />

          <ListMeta
            actors={movie.actors}
            director={movie.director}
            genres={movie.genres}
          />

          <Spacer small={editOpen} />

          <Location
            fileSize={movie.fileSize}
            location={movie.location}
            handleClick={this.showMovieInFinder}
            fileExists={this.state.fileAvailable}
          />

          {editOpen &&
            <Update
              onRedoSearch={this.handleRedoSearch}
              onSaveSearch={searchMovieResult ? this.handleSaveSearch : undefined}
              searchQuery={movie.query}
              searching={searching}
              searchError={searchError}
            />
          }
        </MovieDetailsContainer>

      </FlexboxDiv>
    )
  }
}

const FlexboxDiv = styled.div`
  display: flex;
  margin-top: 20px;
`

const EditButton = styled(Edit)`
  color: rgba(255, 255, 255, 0.9);
  cursor: pointer;
  font-size: 30px;
  padding: 5px 0 0;
  margin-left: 24px;
`

const EditButtonRed = styled(EditButton)`
  color: red;
`

const TrashButton = styled(Trash)`
  color: rgba(255, 255, 255, 0.9);
  cursor: pointer;
  font-size: 30px;
  padding: 5px 0 0;
  margin-left: 24px;
`

const CloseButton = styled(Close)`
  color: rgba(255, 255, 255, 0.9);
  cursor: pointer;
  font-size: 30px;
  padding: 5px 0 0;
  margin-left: 24px;
`

const Poster = styled.aside`
  align-items: center;
  background: ${props => `url(${props.fileUrl}) no-repeat center`};
  display: flex;
  flex-shrink: 0;
  height: 444px;
  justify-content: center;
  margin-right: 20px;
  width: 300px;
  ${props => !props.fileUrl && `
    border: 1px solid #999;
  `}
`

const MovieDetailsContainer = styled.article`
  color: rgba(255, 255, 255, 0.7);
  display: flex;
  flex-direction: column;
  height: 444px;
  min-width: 500px;
  max-width: 600px;
`

const VerticalScrollSection = styled.section`
  overflow-y: auto;
`

const Spacer = styled.div`
  margin-top: ${props => props.small ? '10px' : '30px'};
`
