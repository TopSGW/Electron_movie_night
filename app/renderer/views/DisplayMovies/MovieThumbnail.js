import { shell } from 'electron'
import React, { Component } from 'react'
import styled from 'styled-components'
import { fileExists } from '../../../shared/utils'
import { Angle } from '../../icons'
import { fadeIn, fadeOut } from '../styleUtils'
import PlayButton from './MovieDetail/PlayMovieButton'

export default class MovieThumbnail extends Component {
  constructor (props) {
    super(props)
    this.state = {
      controls: 'hide', // hide, fadingIn, fadingOut, show
      image: 'show', // show, fadeOut
      fileAvailable: true
    }

    this.mouseEnter = this.mouseEnter.bind(this)
    this.mouseLeave = this.mouseLeave.bind(this)
    this.onAnimationEnd = this.onAnimationEnd.bind(this)

    this.openMovieInDefaultPlayer = this.openMovieInDefaultPlayer.bind(this)
    this.onShowMovieDetailsClick = this.onShowMovieDetailsClick.bind(this)
    this.renderControls = this.renderControls.bind(this)
  }

  mouseEnter (e) {
    e.preventDefault()
    this.setState({ controls: 'fadingIn' })
    fileExists(this.props.movie.location)
      .then(result => this.setState({ fileAvailable: result }))
  }

  mouseLeave (e) {
    e.preventDefault()
    if (this.state.controls === 'show') {
      this.setState({ controls: 'fadingOut' })
    } else {
      this.setState({ controls: 'hide' })
    }
  }

  onAnimationEnd (e) {
    e.preventDefault()
    if (this.state.controls === 'fadingIn') {
      this.setState({ controls: 'show' })
    } else if (this.state.controls === 'fadingOut') {
      this.setState({ controls: 'hide' })
    }
  }

  openMovieInDefaultPlayer (e) {
    e.preventDefault()
    shell.openItem(this.props.movie.location)
  }

  onShowMovieDetailsClick (e) {
    e.preventDefault()
    const { movie, handleMovieSelected, panelID } = this.props
    if (handleMovieSelected) {
      handleMovieSelected({ movie, action: 'click', panelID })
    }
  }

  render () {
    return (
      <ImageContainer
        onMouseEnter={this.mouseEnter}
        onMouseLeave={this.mouseLeave}
      >
        {this.renderImage()}
        {this.renderControls()}
      </ImageContainer>
    )
  }

  renderImage () {
    const { imgFile, title } = this.props.movie

    if (!imgFile) {
      return <NoImage>{title}</NoImage>
    }

    return <Image
      src={imgFile}
      alt={title}
    />
  }

  renderControls () {
    if (this.state.controls === 'hide') {
      return null
    }

    const Tag = this.state.controls === 'fadingOut' ? HoverOut : HoverIn

    return (
      <Tag onAnimationEnd={this.onAnimationEnd} >
        {this.state.fileAvailable && <PlayButton onClick={this.openMovieInDefaultPlayer} small />}
        <ShowMovieDetailsArrow
          down
          onClick={this.onShowMovieDetailsClick}
        />
      </Tag>
    )
  }
}

const ImageContainer = styled.div`
  margin: 5px 10px 5px 0;
  background-color: #141414;
  color: #999;
  display: flex;
  position: relative;
`

const Image = styled.img`
  width: 150px;
  height: 222px;
`

const NoImage = styled.div`
  width: 150px;
  height: 222px;
  border: 1px solid #999;
  padding: 8px;
  overflow-wrap: break-word;
`

const Hover = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`

const HoverIn = Hover.extend`
  animation: 0.5s ${fadeIn} ease-in;
`

const HoverOut = Hover.extend`
  animation: 0.3s ${fadeOut} ease-out;
`

const ShowMovieDetailsArrow = styled(Angle)`
  color: rgba(255, 255, 255, 0.9);
  cursor: pointer;
  position: absolute;
  bottom: 0;
  font-size: 50px;
  margin-bottom: -5px;
  overflow: hidden;
  text-align: center;
  text-shadow: 0 0 1px #141414;
  transition: all 0.2s ease-in-out;
  &:hover {
    color: rgba(2, 117, 216, 1);
  }
`
