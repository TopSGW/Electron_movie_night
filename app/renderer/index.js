import React from 'react'
import ReactDOM from 'react-dom'
import { Provider } from 'react-redux'

import { logEnv } from '../shared/utils'
import { App } from './controller'
import store from './model'
import logger from './mainWindowLogger'

logEnv(logger)

// Render the App.
ReactDOM.render(
  <Provider store={store}>
    <App />
  </Provider>,
  document.getElementById('root')
)
