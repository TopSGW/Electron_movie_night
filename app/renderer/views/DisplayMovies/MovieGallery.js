import React, { Component } from 'react'
import styled from 'styled-components'
import { Chevron } from '../../icons'
import { MovieThumbnailContainer } from '../../controller'

import { fadeIn } from '../styleUtils'

export default class MovieGallery extends Component {
  constructor (props) {
    super(props)
    this.selectCategory = this.selectCategory.bind(this)
  }

  selectCategory (category) {
    return (e) => {
      e.preventDefault()
      const { handleSelectCategory } = this.props
      if (handleSelectCategory) {
        handleSelectCategory(category)
      }
    }
  }

  render () {
    const { category } = this.props

    return (
      <Gallery key={category}>
        { this.renderTitle() }
        { this.renderThumbnails() }
      </Gallery>
    )
  }

  renderTitle () {
    const { category, renderStyle } = this.props
    if (!category) {
      return null
    }

    if (renderStyle === 'grid') {
      return <Title>{category}</Title>
    }

    return (
      <Title>
        <Genre onClick={this.selectCategory(category)}>
          {category}
          <Chevron right fixedWidth />
        </Genre>
      </Title>
    )
  }

  renderThumbnails () {
    const { movies, renderStyle, id } = this.props

    const thumbnails = movies.map((movie) => {
      return (
        <FadeIn key={movie.location}>
          <MovieThumbnailContainer
            // key={movie.location}
            movie={movie}
            panelID={id}
          />
        </FadeIn>
      )
    })

    if (renderStyle === 'grid') {
      return <GridContainer>{thumbnails}</GridContainer>
    } else {
      return <HorizontalScrollContainer>{thumbnails}</HorizontalScrollContainer>
    }
  }
}

const FadeIn = styled.div`
  animation: 0.5s ${fadeIn} ease-in;
`

const Gallery = styled.div`
  margin-top: 30px;
  margin-right: 10px;
`

const Title = styled.h2`
  line-height: 1.3;
  margin: 0;
  display: block;
  font-size: 1.2em;
  font-weight: bold;
  color: #999;
  font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
  background-color: transparent;
  text-transform: capitalize;
`

const Genre = styled.span`
  cursor: pointer;
  transition: all 0.2s ease-in-out;
  &:hover {
    color: rgba(255, 255, 255, 0.7);
  }
`

const HorizontalScrollContainer = styled.div`
  display: flex;
  overflow-x: auto;
  &::-webkit-scrollbar {
    height: 12px;
  }

  &::-webkit-scrollbar-button {
    display: none;
  }

  &::-webkit-scrollbar-track {
    display: none;
  }

  &::-webkit-scrollbar-thumb {
    display: block;
    border-radius: 10px;
    background-color: rgba(255, 255, 255, 0.1);
  }
`

const GridContainer = styled.div`
  display: flex;
`
