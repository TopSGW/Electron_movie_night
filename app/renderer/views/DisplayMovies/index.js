import React, { Component } from 'react'
import styled from 'styled-components'
import { gridPartition } from '../../../shared/utils'
import MovieDetail from './MovieDetail'
import MovieGallery from './MovieGallery'
import { fadeIn, fadeOut } from '../styleUtils'

export default class DisplayMovies extends Component {
  constructor (props) {
    super(props)
    this.state = { width: 0, closing: false, prevFeaturedMovie: null }

    this.fadeOutAnimationEnded = this.fadeOutAnimationEnded.bind(this)
    this.selectCategory = this.selectCategory.bind(this)
    this.updateWindowDimensions = this.updateWindowDimensions.bind(this)
  }

  // React Lifecycle Method.
  // Based on current props and incoming props, decide if we can
  // fade out the currently displayed movie details component.
  componentWillReceiveProps (nextProps) {
    if (DisplayMovies.shouldFadeOutMovieDetails(this.props, nextProps)) {
      this.setState({ closing: true, prevFeaturedMovie: this.props.featuredMovie })
    }
  }

  componentDidMount () {
    this.updateWindowDimensions()
    window.addEventListener('resize', this.updateWindowDimensions)
  }

  componentWillUnmount () {
    window.removeEventListener('resize', this.updateWindowDimensions)
  }

  fadeOutAnimationEnded (e) {
    e.preventDefault()
    this.setState({ closing: false, prevFeaturedMovie: null })
  }

  updateWindowDimensions () {
    this.setState({ width: window.innerWidth })
  }

  selectCategory (category) {
    const { updateSearchCategory } = this.props
    if (updateSearchCategory) {
      updateSearchCategory(category)
    }
  }

  render () {
    const { featuredMovie, movies, clearSearchQuery } = this.props

    // Handle no search results
    if (movies.length === 0) {
      return <NoMovieStyles>No Matches</NoMovieStyles>
    }

    // Special handling for featured movie from search result
    const { closing, prevFeaturedMovie } = this.state
    if (featuredMovie && featuredMovie.action === 'search') {
      const onClose = () => clearSearchQuery()
      return this.renderMovieDetails(onClose)
    } else if (closing && prevFeaturedMovie && prevFeaturedMovie.action === 'search') {
      return this.renderMovieDetails() // fade out search result
    }

    // If we only have one genre, then split it out over multiple
    // fixed length movie gallery rows. I.e. create one vertical scroll
    // browser instead of a single horizontal scroll browser.
    if (movies.length === 1) {
      const genre = movies[0].genre
      const allMovies = movies[0].movies

      const imagesPerPartition = Math.floor(this.state.width / 153) || 1
      const grid = gridPartition(allMovies, imagesPerPartition)
      let finalMovies = grid.map((movieList, index) => {
        return index === 0
          ? { genre, movies: movieList }
          : { movies: movieList }
      })

      return (
        <MoviesGroupedByGenre>
          {
            finalMovies.map((item, index) => {
              return this.renderGallery(item, index, 'grid')
            })
          }
        </MoviesGroupedByGenre>
      )
    }

    // Render the movie galleries.
    return (
      <MoviesGroupedByGenre>
        { movies.map((item, index) => this.renderGallery(item, index)) }
      </MoviesGroupedByGenre>
    )
  }

  renderGallery (item, index, renderStyle = 'row') {
    const { genre, movies } = item
    const { featuredMovie, clearFeaturedMovie } = this.props
    const { prevFeaturedMovie } = this.state
    const match = prevFeaturedMovie
      ? prevFeaturedMovie.panelID
      : featuredMovie.panelID

    return (
      <div key={index}>
        <MovieGallery
          key={index}
          category={genre}
          movies={movies}
          id={index}
          renderStyle={renderStyle}
          handleSelectCategory={this.selectCategory}
        />
        { index === match && this.renderMovieDetails(clearFeaturedMovie) }
      </div>
    )
  }

  renderMovieDetails (onClose) {
    const { deleteMovieFromDb, isCrawling, updateMovieMetadata } = this.props
    const { closing, prevFeaturedMovie } = this.state
    if (closing) {
      return (
        <FadeOut
          key={`out-${prevFeaturedMovie.location}`}
          onAnimationEnd={this.fadeOutAnimationEnded}
        >
          <MovieDetail movie={prevFeaturedMovie.movie} />
        </FadeOut>
      )
    }

    const { featuredMovie } = this.props
    const movie = featuredMovie.movie
    if (!movie) {
      return null
    }

    return (
      <FadeIn key={movie.location}>
        <MovieDetail
          allowEdit={!isCrawling}
          movie={movie}
          handleCloseMovieDetails={onClose}
          center={featuredMovie.action === 'click'}
          onMoveToTrash={deleteMovieFromDb}
          onUpdateMovieMetadata={updateMovieMetadata}
        />
      </FadeIn>
    )
  }

  // Helper function for determining whether or not we
  // should fade out the MovieDetails component.
  // Exported here for testing purposes.
  static shouldFadeOutMovieDetails (currProps, nextProps) {
    // Handle cases when we should never animate.
    const curr = currProps.featuredMovie
    if (!curr.movie) { return false } // nothing to fade out

    if (nextProps.searchCategory !== currProps.searchCategory) {
      return false // do not animate when the search category changes
    }

    // Handle case when there is a new incoming featured movie
    const next = nextProps.featuredMovie
    if (next.movie) {
      if (next.action === 'search' ||
          next.movie.location === curr.movie.location ||
          (next.panelID >= 0 && next.panelID !== curr.panelID)) {
        return false
      }
    }

    // Handle case where there is no incoming featured movie
    if (!next.movie) {
      if (curr.action === 'search' && nextProps.movies.length === 0) {
        return false
      }
    }

    return true
  }
}

const MoviesGroupedByGenre = styled.div`
  padding-left: 20px;
`

const FadeIn = styled.div`
  animation: 0.5s ${fadeIn} ease-in;
`

const FadeOut = styled.div`
  animation: 0.3s ${fadeOut} ease-out;
`
const NoMovieStyles = styled.div`
  align-items: center;
  background-color: rgba(20, 20, 20, 1);
  bottom: 0;
  color: rgba(2, 117, 216, 1);
  display: flex;
  flex-direction: column;
  font-size: 3rem;
  font-family: CopperPlate, serif;
  justify-content: center;
  left: 0;
  padding-bottom: 20%;
  position: fixed;
  right: 0;
  top: 50px;
`
