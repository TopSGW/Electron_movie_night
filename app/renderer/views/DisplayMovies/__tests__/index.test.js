'use strict'
/* globals jest, describe, test, expect */

import React from 'react'
import renderer from 'react-test-renderer'

// { searchCategory, updateSearchCategory, clearSearchQuery
// featuredMovie, clearFeaturedMovie
// movies }
import DisplayMovies from '../index'

jest.mock('../MovieGallery', () => {
  return (props) => <div {...props}>MovieGalleryMock</div>
})

jest.mock('../MovieDetail', () => {
  return (props) => <div {...props}>MovieDetailMock</div>
})

const movie = {
  actors: ['Ben Stiller', 'Will Ferrell'],
  director: 'Todd Phillips',
  imgFile: '/path/to/img',
  fileInfo: [
    { location: '/path/to/movie' }
  ],
  genres: ['action', 'comedy', 'drama'],
  imdbID: 'tt123'
}

const movieDB = [
  { genre: 'action', movies: [movie] }
]

describe('DisplayMovies', () => {
  test('renders No Matches when no movies', () => {
    const component = renderer.create(
      <DisplayMovies
        movies={[]}
      />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    expect(tree.children[0]).toEqual('No Matches')
  })

  test('renders featured movie from search result', () => {
    const component = renderer.create(
      <DisplayMovies
        movies={movieDB}
        featuredMovie={{ movie, action: 'search', panelID: -1 }}
      />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('renders single movie using grid style', () => {
    const component = renderer.create(
      <DisplayMovies
        movies={movieDB}
        featuredMovie={{movie: null, action: '', panelID: -1}}
      />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('renders multiple movies using row style', () => {
    let testDB = JSON.parse(JSON.stringify(movieDB))
    testDB.push({ genre: 'drama', movies: [movie] })

    const component = renderer.create(
      <DisplayMovies
        movies={testDB}
        featuredMovie={{movie: null, action: '', panelID: 0}}
      />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('clicking on category title executes updateSearchCategory callback', () => {
    const handleUpdate = jest.fn()

    const component = renderer.create(
      <DisplayMovies
        movies={movieDB}
        featuredMovie={{movie: null, action: '', panelID: -1}}
        updateSearchCategory={handleUpdate}
      />
    )
    let tree = component.toJSON()
    const movieGallery = tree.children[0].children[0]
    movieGallery.props.handleSelectCategory('action')
    expect(handleUpdate).toHaveBeenLastCalledWith('action')
  })

  test('animates featured movie', () => {
    // render component with featured movie
    const featured = { movie, action: 'search', panelID: -1 }
    let component = renderer.create(
      <DisplayMovies
        movies={movieDB}
        featuredMovie={featured}
      />
    )

    // update component with no featured movie this time.
    // this will render a fading out featured movie
    component.update(
      <DisplayMovies
        movies={movieDB}
        featuredMovie={{ movie: null, action: '', panelID: -1 }}
      />
    )

    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    // Simulate animation end event causing component
    // to re-render with a movie gallery.
    tree.props.onAnimationEnd({ preventDefault: jest.fn() })
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    // Unmount the component
    component.unmount()
  })

  test('shouldFadeOutMovieDetails', () => {
    const fm1 = { movie, action: 'search', panelID: -1 }
    const fm2 = { movie: null, action: '', panelID: -1 }

    let movie2 = JSON.parse(JSON.stringify(movie))
    movie2.imdbID = 'tt456'
    let fm3 = { movie: movie2, action: 'click', panelID: 0 }

    // nothing to fade out returns false
    let curr = { featuredMovie: fm2 }
    let result = DisplayMovies.shouldFadeOutMovieDetails(curr, {})
    expect(result).toBe(false)

    // search category changes returns false
    curr = { featuredMovie: fm1, searchCategory: 'action' }
    let next = { featuredMovie: fm2, searchCategory: '' }
    result = DisplayMovies.shouldFadeOutMovieDetails(curr, next)
    expect(result).toBe(false)

    // new incoming featured movie
    curr = { featuredMovie: fm1 }
    next = { featuredMovie: fm3 }
    result = DisplayMovies.shouldFadeOutMovieDetails(curr, next)
    expect(result).toBe(false)

    // no incoming movies
    curr = { featuredMovie: fm1 }
    next = { featuredMovie: fm2, movies: [] }
    result = DisplayMovies.shouldFadeOutMovieDetails(curr, next)
    expect(result).toBe(false)

    // closing featured movie typical case
    curr = { featuredMovie: fm3 }
    next = { featuredMovie: fm2 }
    result = DisplayMovies.shouldFadeOutMovieDetails(curr, next)
    expect(result).toBe(true)
  })
})
