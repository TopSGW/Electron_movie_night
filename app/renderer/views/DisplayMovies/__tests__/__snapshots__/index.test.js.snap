// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplayMovies animates featured movie 1`] = `
<div
  className="sc-htpNat fwfEAa"
  onAnimationEnd={[Function]}
>
  <div
    movie={
      Object {
        "actors": Array [
          "Ben Stiller",
          "Will Ferrell",
        ],
        "director": "Todd Phillips",
        "fileInfo": Array [
          Object {
            "location": "/path/to/movie",
          },
        ],
        "genres": Array [
          "action",
          "comedy",
          "drama",
        ],
        "imdbID": "tt123",
        "imgFile": "/path/to/img",
      }
    }
  >
    MovieDetailMock
  </div>
</div>
`;

exports[`DisplayMovies animates featured movie 2`] = `
<div
  className="sc-bdVaJa gECVQw"
>
  <div>
    <div
      category="action"
      handleSelectCategory={[Function]}
      id={0}
      movies={
        Array [
          Object {
            "actors": Array [
              "Ben Stiller",
              "Will Ferrell",
            ],
            "director": "Todd Phillips",
            "fileInfo": Array [
              Object {
                "location": "/path/to/movie",
              },
            ],
            "genres": Array [
              "action",
              "comedy",
              "drama",
            ],
            "imdbID": "tt123",
            "imgFile": "/path/to/img",
          },
        ]
      }
      renderStyle="grid"
    >
      MovieGalleryMock
    </div>
  </div>
</div>
`;

exports[`DisplayMovies renders No Matches when no movies 1`] = `
<div
  className="sc-bxivhb cFqVw"
>
  No Matches
</div>
`;

exports[`DisplayMovies renders featured movie from search result 1`] = `
<div
  className="sc-bwzfXH gNLDG"
>
  <div
    center={false}
    handleCloseMovieDetails={[Function]}
    movie={
      Object {
        "actors": Array [
          "Ben Stiller",
          "Will Ferrell",
        ],
        "director": "Todd Phillips",
        "fileInfo": Array [
          Object {
            "location": "/path/to/movie",
          },
        ],
        "genres": Array [
          "action",
          "comedy",
          "drama",
        ],
        "imdbID": "tt123",
        "imgFile": "/path/to/img",
      }
    }
    onMoveToTrash={undefined}
    onUpdateMovieMetadata={undefined}
  >
    MovieDetailMock
  </div>
</div>
`;

exports[`DisplayMovies renders multiple movies using row style 1`] = `
<div
  className="sc-bdVaJa gECVQw"
>
  <div>
    <div
      category="action"
      handleSelectCategory={[Function]}
      id={0}
      movies={
        Array [
          Object {
            "actors": Array [
              "Ben Stiller",
              "Will Ferrell",
            ],
            "director": "Todd Phillips",
            "fileInfo": Array [
              Object {
                "location": "/path/to/movie",
              },
            ],
            "genres": Array [
              "action",
              "comedy",
              "drama",
            ],
            "imdbID": "tt123",
            "imgFile": "/path/to/img",
          },
        ]
      }
      renderStyle="row"
    >
      MovieGalleryMock
    </div>
  </div>
  <div>
    <div
      category="drama"
      handleSelectCategory={[Function]}
      id={1}
      movies={
        Array [
          Object {
            "actors": Array [
              "Ben Stiller",
              "Will Ferrell",
            ],
            "director": "Todd Phillips",
            "fileInfo": Array [
              Object {
                "location": "/path/to/movie",
              },
            ],
            "genres": Array [
              "action",
              "comedy",
              "drama",
            ],
            "imdbID": "tt123",
            "imgFile": "/path/to/img",
          },
        ]
      }
      renderStyle="row"
    >
      MovieGalleryMock
    </div>
  </div>
</div>
`;

exports[`DisplayMovies renders single movie using grid style 1`] = `
<div
  className="sc-bdVaJa gECVQw"
>
  <div>
    <div
      category="action"
      handleSelectCategory={[Function]}
      id={0}
      movies={
        Array [
          Object {
            "actors": Array [
              "Ben Stiller",
              "Will Ferrell",
            ],
            "director": "Todd Phillips",
            "fileInfo": Array [
              Object {
                "location": "/path/to/movie",
              },
            ],
            "genres": Array [
              "action",
              "comedy",
              "drama",
            ],
            "imdbID": "tt123",
            "imgFile": "/path/to/img",
          },
        ]
      }
      renderStyle="grid"
    >
      MovieGalleryMock
    </div>
  </div>
</div>
`;
